### ISSUE TITLE
Why only 2 scale Sinkhorn?

### ISSUE THREAD

**Issue by SamuelJoutard** [jeanfeydy/geomloss#8]
Hi Jean,

Very nice work congratulations!

I am quite new to optimal transport but I was wondering why you only implemented 2 scale Sinkhorn algorithm for the multi-scale approach? It seems to me that just adding a for loop with the number of loop being computed on the fly based on input parameters or distribution characteristic would not change much and could have some benefits. Maybe it is actually more complex to propagate the approximate solution over multiple scales? 

Thank you in advance for your help.

Best,

Samuel

**Comment by jeanfeydy** [comment]
Hi Samuel,

Good question, thanks for asking it!
The tl;dr is that it is not worth the trouble for distributions with less than 200k samples each. And as of today, I don't know if anyone really needs to go beyond that.

If we profile the `SampleLoss(backend="multiscale")` layer as described [here](https://www.kernel-operations.io/geomloss/_auto_examples/performances/plot_profile.html), we see that the computation time is generally dominated at 80%-90% by the iterations at the finest scale: in a geometric series, the last term is always preponderant. Consequently, performing the first few iterations on a multiscale decomposition would not yield any sensible speed-up.
In the other direction, refining the current clustering scheme (which typically puts samples in 500-1,000 cubic boxes) to increase the number of clusters in a multiscale fashion could help us to scale up to larger point clouds: as evidenced by the last benchmark of [this page](https://www.kernel-operations.io/geomloss/_auto_examples/performances/plot_benchmarks_samplesloss_3D.html), the one and two-scales GeomLoss routines start to leave the sweet `O(n*log(n))` regime for the dreaded `O(n^2)` around `n = 200k`. But this would require some low-level work on the [kernel truncation scheme ](https://github.com/jeanfeydy/geomloss/blob/0a8329b20c939e8c937b3cfd9af6c43d0481c555/geomloss/sinkhorn_samples.py#L156), which is currently implemented using the (not-so-efficient) advanced indexing of PyTorch on a large Nclusters-by-Mclusters boolean matrix, in the [`from_matrix`](https://github.com/getkeops/keops/blob/a6cd74d27d1e7c4be8b24746fd265bb90ee50175/pykeops/torch/cluster/matrix.py#L112) routine. As of today, it will break down if Nclusters * Mclusters exceeds something like 10,000,000.


Note that I agree with you: implementing a fully multiscale approach could be useful if we really wanted to scale to super-large point clouds. If you inspect the multiscale Sinkhorn code, you will see that it is, in fact, already implemented this way. But would that really be useful to anybody? In most applied settings (e.g. in medical imaging), sampling a 3D mesh with 200k triangles is already more than good enough: our registration models are not precise or robust enough to handle a higher level of accuracy anyway.

As far as I can tell, the current GeomLoss interface provides a decent trade-off between simplicity and usability. The fact that users can specify their own clustering scheme using a simple vector of labels is a [nice feature](https://www.kernel-operations.io/geomloss/_auto_examples/sinkhorn_multiscale/plot_optimal_transport_cluster.html) I think. 
But I'm open to interactions! Needless to say, I'd be happy to help or work on it if you believe that your application would actually benefit from it.
What do you think?

Best regards,

Jean

P.S.: By the way, as I see that you're from King's: I'm starting a PostDoc at Imperial College London this week (for 2-3 years), and would be more than happy to pay you a visit anytime :-)

**Comment by SamuelJoutard** [comment]
Hi Jean,

Thank you very much for you detailed answer! I understand what you mean and I agree with you that SamplesLoss in its current form has lost of interesting features and is very convenient. 

I was asking that because I would like to start from you code to implement some ideas I had that would imply some additional steps at each scale, and it would be convenient to have more scales to make this additional processing easier (and if it works well I could contribute to GeomLoss if possible).

I would be glad to meet you to talk about all this since you are in London as well now! I will get in touch by email to give you more details (@ens.fr mail address).

Best regards,

Samuel
